tools/preconf-rpc/sender/sender.go

```diff
@@ -0,0 +1,496 @@
+		return false
+	}
+	if _, ok := t.inflightAccount[txn.Sender]; ok {
+		t.logger.Debug("Transaction sender already has an inflight transaction, skipping", "sender", txn.Sender.Hex())
+		t.triggerSender() // Trigger to reprocess later
+		return false
+	}
+
+	t.inflightTxns[txn.Hash()] = struct{}{}
+	t.inflightAccount[txn.Sender] = struct{}{}
+	return true
+}
+
+func (t *TxSender) markCompleted(txn *Transaction) {
+	t.inflightMu.Lock()
+	defer t.inflightMu.Unlock()
+
+	delete(t.inflightTxns, txn.Hash())
+	delete(t.inflightAccount, txn.Sender)
+}
+
+func (t *TxSender) processQueuedTransactions(ctx context.Context) {
+	txns, err := t.store.GetQueuedTransactions(ctx)
+	if err != nil {
+		t.logger.Error("Failed to get queued transactions", "error", err)
+		return
+	}
+	if len(txns) == 0 {
+		t.logger.Info("No queued transactions to process")
+		return
+	}
+	t.logger.Info("Processing queued transactions", "count", len(txns))
+	for _, txn := range txns {
+		txn := txn // capture range variable
+		select {
+		case <-ctx.Done():
+			t.logger.Info("Context cancelled, stopping transaction processing")
+			return
+		case t.workerPool <- struct{}{}:
+			t.eg.Go(func() error {
+				defer func() { <-t.workerPool }()
+				if !t.markInflight(txn) {
+					// Transaction is already being processed or sender has an inflight transaction
+					return nil
+				}
+				defer t.markCompleted(txn)
+
+				t.logger.Info("Processing transaction", "sender", txn.Sender.Hex(), "type", txn.Type)
+				if err := t.processTransaction(ctx, txn); err != nil {
+					t.logger.Error("Failed to process transaction", "sender", txn.Sender.Hex(), "error", err)
+					txn.Status = TxStatusFailed
+					txn.Details = err.Error()
+					return t.store.StoreTransaction(ctx, txn, nil)
+				}
+				return nil
+			})
+		}
+	}
+}
+
+func (t *TxSender) processTransaction(ctx context.Context, txn *Transaction) error {
+	var (
+		result bidResult
+		err    error
+	)
+BID_LOOP:
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		result, err = t.sendBid(ctx, txn)
+		switch {
+		case err != nil:
+			return err
+		case result.optedInSlot:
+			if result.noOfProviders == len(result.commitments) {
+				// This means that all builders have committed to the bid and it
+				// is a primev opted in slot. We can safely proceed to inform the
+				// user that the txn was successfully sent and will be processed
+				txn.Status = TxStatusPreConfirmed
+				txn.BlockNumber = int64(result.blockNumber)
+				t.logger.Info(
+					"Transaction pre-confirmed",
+					"sender", txn.Sender.Hex(),
+					"type", txn.Type,
+					"blockNumber", result.blockNumber,
+					"bidAmount", result.bidAmount.String(),
+				)
+				break BID_LOOP
+			}
+		default:
+		}
+
+		// Wait for block number to be updated to confirm transaction. If failed
+		// we will retry the bid process till user cancels the operation
+		included, err := t.blockTracker.CheckTxnInclusion(ctx, txn.Hash(), result.blockNumber)
```

Comment:
Seems like implementation of `blockTracker.CheckTxnInclusion` could be simplified. Instead of waiting for a block to be sealed, calling `BlockByNumber`, then iterating through every tx, we could just check `eth_getTransactionReceipt`

Reply:
This function is called only in the cases when we dont have enough commitments (if all providers didnt commit for opted in block) or if it is not opted in block and we got some commitments. The function essentially checks only the block which we bid on to see if we can find the transaction. getTransactionReceipt doesnt have block number param afaik. This is still a stand-in. There should potentially be more checks here, for eg querying relay to see if one of the providers that gave us the commitment actually built the block. For now its simply checks for block transactions.